### PR TITLE
Add a Wi-Fi Fixture container

### PR DESCRIPTION
**Pull request by ksperling-apple**
This container runs patched hostapd and wpa_supplicant against the Wi-Fi
interfaces it is given, so tests can drive an AP or a station over the
daemons' control sockets. The patches add the EAP-TLS features Matter
Per-Device Credentials authentication needs.

The container is started for suites that configure radios in the new
network.wifi.interfaces configuration. The container holds them for the
duration of the test suite. Note that the daemons within the container
do not actually interact with those interfaces in any way until a tests
configures them.

Also exclude patches/ and config/ from the spell checker.

Note: The hostapd patches are a subset of the set maintained at
https://github.com/project-chip/matter-openwrt/tree/main/overlay/hostapd/patches.

The python_testing logic to drive hostapd within the fixture is part of https://github.com/project-chip/connectedhomeip/pull/74170

### CODE REVIEW

File: wifi_container/config/hostapd.config [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# CONFIG_CTRL_IFACE_DIR is not exposed as a make variable
+EXTRA_CFLAGS = -DCONFIG_CTRL_IFACE_DIR='"/run/wifi-fixture/hostapd"'
+
+CONFIG_CTRL_IFACE=y
+CONFIG_DRIVER_NL80211=y
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply wifi_container/config/hostapd.config:19]: The docker image builds fine.
I will look into whether adding the LIBNL32 config is useful, but the comment that the image can’t be built is completely wrong. I have been building and running it and exercised a real test script where the fixture provides an AP that an actual DUT joined successfully.

[the remark this answers is not in the document]
Comment (ksperling-apple) [reply wifi_container/config/hostapd.config:19]: Upstream Hostap's drivers.mak automatically runs pkg-config if needed and picks the correct libnl version:
```
ifdef NEED_LIBNL
ifndef CONFIG_LIBNL32
ifndef CONFIG_LIBNL20
ifndef CONFIG_LIBNL_TINY
PKG_CONFIG ?= pkg-config
HAVE_LIBNL3 := $(shell $(PKG_CONFIG) --exists libnl-3.0; echo $$?)
ifeq ($(HAVE_LIBNL3),0)
CONFIG_LIBNL32=y
endif
endif
endif
endif
```

(Note the "0" in HAVE_LIBNL3 is the exit code, not a boolean)

File: test_collections/matter/sdk_tests/support/wifi_container.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+from docker.models.containers import Container
+
+from app.container_manager import container_manager
+from app.singleton import Singleton
+from app.test_engine.logger import test_engine_logger as logger
+from test_collections.matter.test_environment_config import WiFiConfig
+
+# Built from `wifi_container` in this repository.
+# Tracking :latest until the image lands in main and there is a commit to pin.
+DEFAULT_DOCKER_IMAGE = "ghcr.io/project-chip/csa-certification-tool-wifi:latest"
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply test_collections/matter/sdk_tests/support/wifi_container.py:32]: Deliberate, and the comment above the constant says why: the image isn't published yet, so there is no digest to pin. I'm planning to pin it once it lands in main and the first build is pushed. Note that docker_image in WiFiConfig is a debugging override, not the reproducibility mechanism.

Alternatively we can land just the container part first, but then the question would have been what the point of the image is and how it will be used.

File: wifi_container/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        iw iproute2 wireless-regdb \
+    && rm -rf /var/lib/apt/lists/*
+
+# Copy binaries
+COPY --from=builder \
+    /usr/local/sbin/hostapd \
+    /usr/local/sbin/hostapd_cli \
+    /usr/local/sbin/wpa_supplicant \
+    /usr/local/sbin/wpa_cli \
+    /usr/local/sbin
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply wifi_container/Dockerfile]: Changing to /usr/local/sbin/. The image does build without it though -- BuildKit accepts a multi-source COPY into an existing directory. It's a style rule worth satisfying, not a build failure.

File: wifi_container/scripts/build-docker-image.sh [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+
+GHCR_ORG="ghcr.io"
+ORG=${DOCKER_BUILD_ORG:-project-chip}
+
+# Latest commit hash
+GIT_SHA=$(git rev-parse --short HEAD)
+
+# If working copy has changes, append `-local` to hash
+GIT_DIFF=$(git diff -s --exit-code || echo "-local")
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply wifi_container/scripts/build-docker-image.sh:33]: This script is copied from backend/scripts/build-docker-image.sh, and the GIT_DIFF line is unchanged from it, as is the frontend's copy. I'd rather not have the three diverge on this. Note that the script already computes DIRTY=$(git status --porcelain --untracked-files=no) at line 47 and never uses it, so the check you're describing was half-written already. Worth a separate change across all three scripts.

File: wifi_container/scripts/build-docker-image.sh [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+# Don't run `docker build` and `docker image prune` when `--skip-build` in arguments
+[[ ${*/--skip-build//} != "${*}" ]] || {
+    docker build "${BUILD_ARGS[@]}" -t "$GHCR_ORG"/"$ORG"/"$IMAGE":"$VERSION" .
+    docker image prune --force
+}
+
+[[ ${*/--squash//} != "${*}" ]] && {
+    command -v docker-squash >/dev/null &&
+        docker-squash "$GHCR_ORG"/"$ORG"/"$IMAGE":"$VERSION" -t "$GHCR_ORG"/"$ORG"/"$IMAGE":squashed
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply wifi_container/scripts/build-docker-image.sh:98]: The squash and push blocks here are unchanged from backend/scripts/build-docker-image.sh, so same answer as the other build script comment. Also --squash is dead in CI anyway -- build-docker-images.yml passes --squash --push --multiarch, and the --multiarch branch exits before the squash step runs.

This particular container is also only around 80MB, so squashing isn't going to make a practical difference in terms of size.

File: test_collections/matter/sdk_tests/support/wifi_container.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+        # Naming the radios is all the fixture is told. It resolves them itself and
+        # publishes what it managed to claim, so a dongle that failed to enumerate
+        # stops the container here rather than failing a test later on.
+        self.run_parameters["environment"] = [
+            "WIFI_FIXTURE_IFNAMES=" + ",".join(config.interfaces)
+        ]
+
+        if config.docker_image is not None:
+            self.__docker_image = config.docker_image
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply test_collections/matter/sdk_tests/support/wifi_container.py:91]: done

**Comment by ksperling-apple** [comment]
Image build passes in CI: https://github.com/project-chip/certification-tool-backend/actions/runs/35032951972/job/104595431338?pr=370

**Review by rquidute — APPROVED** [review]

File: test_collections/matter/sdk_tests/support/wifi_container.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        """Create the fixture container with the configured radios.
+
+        Returns true once both daemons answer on their global control socket, or
+        false if the container is already running.
+        """
+        if self.is_running():
+            logger.warning(
+                "Wi-Fi fixture container is already running for " + self.__docker_image
+            )
+            return False
```
Comment (oxesoft) [inline test_collections/matter/sdk_tests/support/wifi_container.py:112]: Not addressed by the discussion so far: when `is_running()` is `True` here, `start()` returns `False` without ever calling `__load_config`, and both call sites (`PythonTestSuite.setup()`) discard the return value. So a leftover fixture container silently keeps whatever interfaces/`docker_image` a *previous* suite configured, and the current suite gets no error at all.

That state is reachable in practice: `self.__container` only goes back to `None` via `destroy()`, which is the last line of `PythonTestSuite.cleanup()`, after `sdk_container.destroy()` and `border_router.destroy_device()` — neither guarded by try/except (unlike the admin_storage capture step right above them, which is explicitly wrapped "so don't block container/border-router teardown"). Both of those do real docker calls (`exec`, container removal) that can genuinely fail transiently. If either raises, `wifi_container.destroy()` never runs, and the fixture container/singleton state leaks into the next suite that requests it.

Since Test Harness backends normally stay up across multiple sequential runs, this isn't just a unit-test corner case. And because it fails silently rather than erroring, the DUT could end up exercised against a stale AP/station config with no indication to the operator — which is a worse failure mode for a certification tool than a loud crash, since it can produce a misleading pass/fail verdict instead of an obvious error.

Suggest: raise (or otherwise surface) a `WiFiContainerError` here instead of silently returning `False`, and/or guard the `sdk_container`/`border_router` teardown steps in `cleanup()` so `wifi_container.destroy()` is guaranteed to run even if an earlier step throws.

File: wifi_container/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+COPY config/hostapd.config /src/hostap/hostapd/.config
+COPY config/wpa_supplicant.config /src/hostap/wpa_supplicant/.config
+COPY patches/*.patch /src/hostap/patches/
+
+RUN cd hostap \
+    && for p in patches/*.patch; do patch -p1 <"$p" || exit 1; done \
+    && make -j"$(nproc)" install CONFIG_NO_GITVER=y BINDIR=/usr/local/sbin
+
+# ---- runtime ----------------------------------------------------------------
+FROM ubuntu:22.04 AS runtime
```
Comment (greens) [inline wifi_container/Dockerfile:42]: The other TH images are also on 22.04, but they should probably be updated to 24.04, the actual target platform for the TH. The sdk image is correctly on 24.04 already.

(not that it makes a huge difference)

File: test_collections/matter/sdk_tests/support/wifi_container.py (hunk 7 above)
Comment (oxesoft) [reply test_collections/matter/sdk_tests/support/wifi_container.py:112]: Filed as #373 to track separately.

File: wifi_container/Dockerfile (hunk 8 above)
Comment (oxesoft) [reply wifi_container/Dockerfile:42]: Filed as #374 to track separately (aligning all TH images on 24.04).
